x/wasm/keeper/handler_plugin.go

```diff
@@ -161,54 +163,115 @@ func (m MessageHandlerChain) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAd
+		}
+		moduleLogger(ctx).Debug("ibc packet set", "seq", seq)
 
-	resp := &types.MsgIBCSendResponse{Sequence: seq}
-	val, err := resp.Marshal()
-	if err != nil {
-		return nil, nil, nil, errorsmod.Wrap(err, "failed to marshal IBC send response")
-	}
-	any, err := codectypes.NewAnyWithValue(resp)
-	if err != nil {
-		return nil, nil, nil, errorsmod.Wrap(err, "failed to convert IBC send response to Any")
+		resp := &types.MsgIBCSendResponse{Sequence: seq}
+		val, err := resp.Marshal()
+		if err != nil {
+			return nil, nil, nil, errorsmod.Wrap(err, "failed to marshal IBC send response")
+		}
+		any, err := codectypes.NewAnyWithValue(resp)
+		if err != nil {
+			return nil, nil, nil, errorsmod.Wrap(err, "failed to convert IBC send response to Any")
+		}
+		msgResponses := [][]*codectypes.Any{{any}}
+
+		return nil, [][]byte{val}, msgResponses, nil
+	case msg.IBC.WriteAcknowledgement != nil:
+		if contractIBCPortID == "" {
+			return nil, nil, nil, errorsmod.Wrapf(types.ErrUnsupportedForContract, "ibc not supported")
+		}
+		contractIBCChannelID := msg.IBC.WriteAcknowledgement.ChannelID
+		if contractIBCChannelID == "" {
+			return nil, nil, nil, errorsmod.Wrapf(types.ErrEmpty, "ibc channel")
+		}
+
+		channelCap, ok := h.capabilityKeeper.GetCapability(ctx, host.ChannelCapabilityPath(contractIBCPortID, contractIBCChannelID))
+		if !ok {
+			return nil, nil, nil, errorsmod.Wrap(channeltypes.ErrChannelCapabilityNotFound, "module does not own channel capability")
+		}
+
+		packet, err := h.wasmKeeper.LoadAsyncAckPacket(ctx, contractIBCPortID, contractIBCChannelID, msg.IBC.WriteAcknowledgement.PacketSequence)
+		if err != nil {
+			return nil, nil, nil, errorsmod.Wrap(types.ErrInvalid, "packet")
+		}
+
+		err = h.ics4Wrapper.WriteAcknowledgement(ctx, channelCap, packet, asyncAck(msg.IBC.WriteAcknowledgement.Ack))
+		if err != nil {
+			return nil, nil, nil, errorsmod.Wrap(err, "acknowledgement")
+		}
+
+		// delete the packet from the store after acknowledgement
+		h.wasmKeeper.DeleteAsyncAckPacket(ctx, contractIBCPortID, contractIBCChannelID, msg.IBC.WriteAcknowledgement.PacketSequence)
+
+		resp := &types.MsgIBCWriteAcknowledgementResponse{}
+		val, err := resp.Marshal()
+		if err != nil {
+			return nil, nil, nil, errorsmod.Wrap(err, "failed to marshal IBC send response")
+		}
+
+		any, err := codectypes.NewAnyWithValue(resp)
+		if err != nil {
+			return nil, nil, nil, errorsmod.Wrap(err, "failed to convert IBC send response to Any")
+		}
+		msgResponses := [][]*codectypes.Any{{any}}
+
+		return nil, [][]byte{val}, msgResponses, nil
+	default:
+		return nil, nil, nil, types.ErrUnknownMsg
 	}
-	msgResponses := [][]*codectypes.Any{{any}}
+}
+
+var _ ibcexported.Acknowledgement = asyncAck{}
+
+type asyncAck wasmvmtypes.IBCAcknowledgement
+
+func (w asyncAck) Success() bool {
+	// We don't give the contract to supply a success flag because it's not part of the IBC spec.
+	// Currently, ibc-go also just ignores this flag for `WriteAcknowledgement`.
```

Comment:
I believe this is ignored when handling an acknowledgement asynchronously, which is the case here. However, this is used if the acknowledgement is synchronous, see here. We discard the state changes due to application callbacks if this is false. Since the entire contract state is in the application callback, even if the ack was synchronous, you'd want to always return true. Right?

Reply:
Yes, the success value is ignored for async acks. This type is specifically for async acks. For synchronous acks, there is a separate type `ContractConfirmStateAck`, which also just returns `true`, so I guess that could be reused here instead.